{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 556,
   "id": "8aef3002",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from geopy.distance import geodesic\n",
    "import networkx as nx \n",
    "from icecream import ic\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "id": "f74ec483",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CITIES = pd.read_csv('cities/italy.csv', header=None, names=['name', 'lat', 'lon'])\n",
    "#CITIES = pd.read_csv('cities/china.csv', header=None, names=['name', 'lat', 'lon'])\n",
    "#CITIES = pd.read_csv('cities/russia.csv', header=None, names=['name', 'lat', 'lon'])\n",
    "#CITIES = pd.read_csv('cities/us.csv', header=None, names=['name', 'lat', 'lon'])\n",
    "CITIES = pd.read_csv('cities/vanuatu.csv', header=None, names=['name', 'lat', 'lon'])\n",
    "\n",
    "\n",
    "DISTANCE_MATRIX = np.zeros((len(CITIES), len(CITIES)))\n",
    "\n",
    "for c1, c2 in combinations(CITIES.itertuples(), 2):\n",
    "    DISTANCE_MATRIX[c1.Index, c2.Index] = DISTANCE_MATRIX[c2.Index, c1.Index] = geodesic(\n",
    "        (c1.lat, c1.lon), (c2.lat, c2.lon)\n",
    "    ).km"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "id": "c1807e9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def solution_cost(solution):\n",
    "    assert solution[0] == solution[-1]        # start city = end city\n",
    "    assert set(solution) == set(range(len(CITIES)))        # check if all cities are visited\n",
    "    \n",
    "    cost_tot = 0\n",
    "    for c1, c2 in zip(solution, solution[1:]):\n",
    "        cost_tot += DISTANCE_MATRIX[c1, c2]\n",
    "    return cost_tot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd9d45a",
   "metadata": {},
   "source": [
    "## 1 step lookahead (Greedy)\n",
    "At each step we select a city conseidering the smaller sum of distances between the current city, the new one and the one after the new one.<br>\n",
    "\n",
    "*IDEA and HELP:* **Gabriele Pirilli** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "id": "99972c30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final solution cost: 1475.5280911045297\n",
      "Number of inner iterations 28\n"
     ]
    }
   ],
   "source": [
    "it_counter = 0\n",
    "\n",
    "best_cost = np.inf\n",
    "best_solution = np.nan\n",
    "\n",
    "\n",
    "dist_m = DISTANCE_MATRIX.copy()              # it will be modified\n",
    "visited_cities = np.full(len(CITIES), False)\n",
    "    \n",
    "solution = []\n",
    "start_city = 0\n",
    "solution.append(start_city)\n",
    "\n",
    "city = start_city\n",
    "visited_cities[city] = True\n",
    "  \n",
    "    \n",
    "while not np.all(visited_cities):\n",
    "    \n",
    "    dist_m[:, city] = np.inf\n",
    "    \n",
    "    not_visited_cities = [x for x in range(len(visited_cities)) if not visited_cities[x]]\n",
    "    #ic(not_visited_cities)\n",
    "\n",
    "    best_next_city = None\n",
    "    best_two_step_cost = np.inf\n",
    "\n",
    "    # Find the best next city and its two-step cost\n",
    "    for next_city in not_visited_cities:\n",
    "        it_counter += 1\n",
    "        \n",
    "        dist_m_new = dist_m.copy()  # Copy to modify distances temporarily\n",
    "        dist_m_new[:, next_city] = np.inf\n",
    "\n",
    "        #print(lookahead_cities)\n",
    "        \n",
    "        if len(not_visited_cities) == 1:         # Handling the last city\n",
    "            best_next_city = not_visited_cities[0]\n",
    "            two_step_cost = dist_m[city, next_city] \n",
    "            \n",
    "        else:\n",
    "            lookahead_cities = not_visited_cities.copy()\n",
    "            np.delete(lookahead_cities, np.where(lookahead_cities == next_city))    # We remove the current city\n",
    "            \n",
    "            closest_lookahead_city = min(lookahead_cities, key=lambda x: dist_m_new[next_city, x])\n",
    "            two_step_cost = dist_m[city, next_city] + dist_m_new[next_city, closest_lookahead_city]\n",
    "           \n",
    "            if two_step_cost < best_two_step_cost:\n",
    "                best_two_step_cost = two_step_cost\n",
    "                best_next_city = next_city  \n",
    "        \n",
    "    visited_cities[best_next_city] = True       # Move to the best next city\n",
    "    solution.append(best_next_city)\n",
    "    city = best_next_city       # Update \n",
    "\n",
    "    \n",
    "solution.append(start_city)    # Closing the circle\n",
    "cost = solution_cost(solution)\n",
    "\n",
    "#print(\"Final solution path:\", solution)\n",
    "print(\"Final solution cost:\", cost)\n",
    "print(\"Number of inner iterations\", it_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "999e54ec",
   "metadata": {},
   "source": [
    "## Genetic agorithm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "id": "84e69da6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For reproducibility\n",
    "seed_value = 23\n",
    "np.random.seed(seed_value)\n",
    "random.seed(seed_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e261eb83",
   "metadata": {},
   "source": [
    "Starting point = the result of the Greedy algorithm and its variants. <br><br>\n",
    "The **Hyper-modern GA** was implemented. At each iteration or mutation or crossover was performed. 3 types of mutations (one choosen randomly at each iteration) and 1 type of crossover are available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "id": "1aeb2cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "seed_value = 23\n",
    "np.random.seed(seed_value)\n",
    "random.seed(seed_value)\n",
    "\n",
    "\n",
    "### FUNCTIONS\n",
    "class Individual:\n",
    "    def __init__(self, genome: np.ndarray, fitness: float = None):\n",
    "        self.genome = genome\n",
    "        self.fitness = fitness if fitness is not None else self.calculate_fitness()\n",
    "\n",
    "    def calculate_fitness(self) -> float:\n",
    "        return -float(solution_cost(self.genome))  \n",
    "    \n",
    "    \n",
    "def fitness(individual):      #genome = solution\n",
    "    return -float(solution_cost(individual.genome))\n",
    "\n",
    "\n",
    "\n",
    "### PARENT SELECTIONS:\n",
    "def random_parent_selection(population, n=5):   ### TOURNAMENT SELECTION ###\n",
    "    candidates1 = sorted(np.random.choice(population, n), key=lambda e: e.fitness, reverse=True)   # We choose randomly and then sort by fitness\n",
    "    parent_1 = candidates1[0]   #The best among n selected\n",
    "    \n",
    "    candidates2 = sorted(np.random.choice(population, n), key=lambda e: e.fitness, reverse=True)\n",
    "    parent_2 = candidates2[0]\n",
    "    return parent_1, parent_2\n",
    "\n",
    "\n",
    "def fitness_parent_selection(population):   ### FITNESS-PROPORTIONAL SELECTION ###\n",
    "    total_fitness = sum(ind.fitness for ind in population)      # Total fitness\n",
    "    selection_probs = [ind.fitness / total_fitness for ind in population]   #Selection prob. proportional to fitness\n",
    "    selected_individuals = np.random.choice(population, 2, replace=False, p=selection_probs)   # Selecting 2 parents\n",
    "    return selected_individuals[0], selected_individuals[1]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### CROSSOVERS:\n",
    "def inver_over_crossover(p1: Individual, p2: Individual):   #Function done with Gabriele Pirilli\n",
    "    genome1 = p1.genome\n",
    "    genome2 = p2.genome\n",
    "    \n",
    "    child_genome = [None] * len(genome1)\n",
    "    child_genome[0] = genome1[0]  # Fixed start\n",
    "    child_genome[-1] = genome1[-1]  # Fixed end\n",
    "\n",
    "    # Randomly select two crossover points from the middle of the genome\n",
    "    start_idx = np.random.randint(1, len(genome1) - 1)    # Avoid first and last element\n",
    "    end_idx = np.random.randint(1, len(genome1) - 1)\n",
    "    \n",
    "    if start_idx > end_idx:\n",
    "        start_idx, end_idx = end_idx, start_idx\n",
    "\n",
    "    child_genome[start_idx:end_idx + 1] = genome1[start_idx:end_idx + 1]\n",
    "\n",
    "    current_pos = 1  # Start filling after the first fixed position\n",
    "    for gene in genome2:\n",
    "        if gene not in child_genome:\n",
    "            while current_pos < len(child_genome) - 1 and child_genome[current_pos] is not None:\n",
    "                current_pos += 1\n",
    "            if current_pos < len(child_genome) - 1:  # Ensure we don't overwrite the last fixed element\n",
    "                child_genome[current_pos] = gene\n",
    "\n",
    "    new_individual = Individual(np.array(child_genome))\n",
    "    return new_individual\n",
    "\n",
    "\n",
    "\n",
    "### MUTATIONS:\n",
    "def n_swap_mutation(individual: Individual, n=1):   # Chose 2 positions and swap for n times\n",
    "    mutated_genome = individual.genome.copy()\n",
    "    genome_length = len(mutated_genome)\n",
    "    \n",
    "    for _ in range(n):        \n",
    "        idx1, idx2 = np.random.choice(range(1, genome_length-1), 2, replace=False)      # so we mantain first element = last element \n",
    "        mutated_genome[idx1], mutated_genome[idx2] = mutated_genome[idx2], mutated_genome[idx1]\n",
    "        \n",
    "    new_individual = Individual(mutated_genome)\n",
    "    return new_individual\n",
    "\n",
    "\n",
    "def insert_mutation(individual: Individual):   # Select 2 indices, put them near and the elements in between will be after them\n",
    "    mutated_genome = individual.genome\n",
    "    genome_length = len(mutated_genome)\n",
    "    \n",
    "    idx1, idx2 = np.random.choice(range(1, genome_length-1), 2, replace=False)\n",
    "    \n",
    "    if idx1 > idx2:    # idx1 will be for sure the smaller\n",
    "        idx1, idx2 = idx2, idx1\n",
    "    \n",
    "    element = mutated_genome[idx2]\n",
    "    mutated_genome = np.delete(mutated_genome, idx2)     # Remove idx2 from its original position\n",
    "    mutated_genome = np.insert(mutated_genome, idx1 + 1, element)    # Insert the element right after idx1\n",
    "    \n",
    "    new_individual = Individual(mutated_genome)\n",
    "    return new_individual\n",
    "\n",
    "\n",
    "\n",
    "def inversion_mutation(individual: Individual):     # Select 2 idxs and swap them and all the elements in between\n",
    "    mutated_genome = individual.genome.copy()\n",
    "    genome_length = len(mutated_genome)\n",
    "    \n",
    "    idx1, idx2 = np.random.choice(range(1, genome_length-1), 2, replace=False)\n",
    "    if idx1 > idx2:      # idx1 will be for sure the first\n",
    "        idx1, idx2 = idx2, idx1\n",
    "    mutated_genome[idx1:idx2 + 1] = mutated_genome[idx1:idx2 + 1][::-1]  # Inverting\n",
    "\n",
    "    new_individual = Individual(mutated_genome)\n",
    "    return new_individual"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d57d03e9",
   "metadata": {},
   "source": [
    "##### REMARK: The last element must be always equal to the first one! <br>\n",
    "So we consider range(1, genome_length-1) avoiding to select the 1st and last element in MUTATIONS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 569,
   "id": "6a5dbae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cities: 8\n"
     ]
    }
   ],
   "source": [
    "### PARAMETER DEFINITIOS\n",
    "\n",
    "print('Number of cities:', len(CITIES))\n",
    "\n",
    "nnn = 2   #To define how strong swaps to do, how many mutations at each iteration, how big the population should be\n",
    "div = 2\n",
    "p=[0.9, 0.1]    # More probability of mutation than recombination\n",
    "if len(CITIES)>100:\n",
    "    nnn = 4\n",
    "    div = 4\n",
    "    ppp=[0.8, 0.2]  \n",
    "elif len(CITIES)>300:\n",
    "    nnn = 6\n",
    "    div = 6\n",
    "    ppp=[0.8, 0.2]\n",
    "elif len(CITIES)>500:\n",
    "    nnn = 8\n",
    "    div = 8\n",
    "    ppp=[0.8, 0.2]\n",
    "    \n",
    "POPULATION_SIZE = int(len(CITIES)/div)\n",
    "MAX_GENERATIONS = 100000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 570,
   "id": "4a2ba49d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Greedy:  1475.5280911045297\n"
     ]
    }
   ],
   "source": [
    "### Initializations\n",
    "\n",
    "seed_value = 23\n",
    "np.random.seed(seed_value)\n",
    "random.seed(seed_value)\n",
    "\n",
    "n_it_tot = it_counter         # We already searched for a \"good\" solution\n",
    "initial_individual = Individual(solution)       # Greedy solution\n",
    "print('Greedy: ',solution_cost(initial_individual.genome))\n",
    "\n",
    "population = []\n",
    "population.append(initial_individual)\n",
    "\n",
    "mutation_probabilities = [1/3, 1/3, 1/3]\n",
    "mutations = [n_swap_mutation, insert_mutation, inversion_mutation]\n",
    "crossover = inver_over_crossover\n",
    "\n",
    "champion = max(population, key=lambda individual: individual.fitness)    #Best individual so far\n",
    "best_fitness = champion.fitness\n",
    "n_it_no_change = 0     # \n",
    "\n",
    "for g in range(MAX_GENERATIONS):\n",
    "    n_it_tot = n_it_tot + 1\n",
    "    \n",
    "    if n_it_no_change >= 1000:\n",
    "        break\n",
    "    \n",
    "    \n",
    "    #First step only mutation => Generating additional individuals to create the initial population\n",
    "    if g == 0:   \n",
    "        for it in range(POPULATION_SIZE-1):\n",
    "            new_individual = n_swap_mutation(initial_individual, n=nnn)\n",
    "            population.append(new_individual)\n",
    "            \n",
    "    else:    \n",
    "        # Genetic operators => Randomly choose if to do mutation only (0) or recombination only(1)\n",
    "        choice = np.random.choice([0, 1], p=ppp)   # Higher probability to select MUTATION\n",
    "            \n",
    "        if choice == 0:  #Mutation only\n",
    "            selected_mutation = np.random.choice(mutations, p=mutation_probabilities)  # Choose one mutation considering a probability\n",
    "            selected_individuals = np.random.choice(population, nnn*2, replace=False)     # Select 2 individuals from the population\n",
    "                \n",
    "            for i in selected_individuals:     # Mutate those 2 with the selected method\n",
    "                o = selected_mutation(i)\n",
    "                population.append(o)\n",
    "                \n",
    "        elif choice == 1:   #Recombination only\n",
    "            #Parent selection:\n",
    "            pr1, pr2 = random_parent_selection(population)      # one random parent selection\n",
    "            pf1, pf2 = fitness_parent_selection(population)     # one fitness-proportional selection\n",
    "                                             \n",
    "            o_r = crossover(pr1, pr2)\n",
    "            population.append(o_r)\n",
    "            o_f = crossover(pf1, pf2)\n",
    "            population.append(o_f)\n",
    "\n",
    "  \n",
    "        \n",
    "    population.sort(key=lambda i: i.fitness, reverse=True)\n",
    "    population = population[:POPULATION_SIZE]   # Selection of the best individuals\n",
    "    \n",
    "    current_best = max(population, key=lambda individual: individual.fitness)\n",
    "    if current_best.fitness > best_fitness:\n",
    "        best_fitness = current_best.fitness\n",
    "        n_it_no_change = 0  # Reset stagnant count on improvement\n",
    "    else:\n",
    "        n_it_no_change += 1  # Increment stagnant count if no improvement\n",
    "\n",
    "    champion = current_best\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "id": "08c5174b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genetic:  1345.5449564733105\n",
      "Total it.: 1123\n"
     ]
    }
   ],
   "source": [
    "print('Genetic: ', solution_cost(champion.genome))\n",
    "print('Total it.:', n_it_tot)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
